### DISCUSSION TITLE
How to resolve SyntaxError: Cannot use import statement outside a module when running Vitest and TypeORM

### DISCUSSION THREAD

**Discussion by carlosmoliv** [vitest-dev/vitest#3290] [Q&A]
I found the following error while running Vitest tests that require a database connection:

![pKa0K](https://user-images.githubusercontent.com/52833750/235908170-c195dd18-024e-4a30-95b9-3d77a2e1db20.png)

"SyntaxError: Cannot use import statement outside a module."

This issue appeared after I made some changes to my setup. Here's my new script for typeorm:

"migration:generate": "ts-node -r tsconfig-paths/register ./node_modules/typeorm/cli -d ./src/infra/db/postgres/helpers/data-source.helper.ts migration:generate"

The migration file generated by this script is a TS file named "1683049252085-CreateUsersTable.ts." When I changed the file extension from ".ts" to ".mjs", Vitest started running correctly.

**Comment by carlosmoliv** [comment]
I have already tried solutions from these discussions:

- https://github.com/vitest-dev/vitest/discussions/2751
- https://github.com/vitest-dev/vitest/issues/708#issuecomment-1118628479

And I ended up installing the following dependency: https://github.com/egoist/unplugin-swc, But didn't work for me.

**Comment by sheremet-va** [comment]
Your tsc script generates ESM files and your Node.js script runs them with your package.json. If you don't specify that type of your project is `module`, then it assumes it is CommonJS, and you can't use static imports there. Add `"type": "module"` to your `packge.json` or generate `.mjs` files.

**Reply by carlosmoliv** [reply]
I also have tried to put `type: module` but I got new errors:
- When running vitest: `TypeError: Unknown file extension ".ts" for /home/carlos/projects/codex-solucoes/template-backend-node/src/infra/db/postgres/migrations/1683113898079-CreateUsersTable.ts`

- When running my dev script: `TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /home/carlos/projects/codex-solucoes/template-backend-node/src/main/server.ts`

My Scripts:

```
"test": "vitest --passWithNoTests --silent --run",
"test:integration": "yarn test --no-run -c vitest-integration.config.ts",
"start:dev": "nodemon --watch src -e ts --exec 'ts-node' -r tsconfig-paths/register ./src/main/server.ts",
```

**Reply by carlosmoliv** [reply]
My tsconfig:

```
{
  "compilerOptions": {
    "outDir": "./dist",
    "target": "ES2019",
    "module": "CommonJS",
    "esModuleInterop": true,
    "allowJs": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "removeComments": true,
    "incremental": true,
    "sourceMap": true,
    "skipLibCheck": false,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "resolveJsonModule": true,
    "types": ["vitest/globals"],
    "baseUrl": "./src",
    "paths": {
      "@data/*": ["data/*"],
      "@domain/*": ["domain/*"],
      "@infra/*": ["infra/*"],
      "@main/*": ["main/*"],
      "@presentation/*": ["presentation/*"],
      "@tests/*": ["tests/*"]
    }
  }
}

```

Vitest:

```
import { defineConfig } from 'vitest/config';

export default defineConfig({
  plugins: [
    {
      name: 'a-vitest-plugin-that-changes-config',
      config: () => ({
        test: {
          setupFiles: ['./tests/setup-files/db-global-setup.ts'],
        },
      }),
    },
  ],
  test: {
    include: ['**/*.*-spec.ts'],
    mockReset: true,
    globals: true,
    coverage: {
      reportsDirectory: './coverage',
    },
    root: './',
    alias: {
      '@data': './src/data',
      '@domain': './src/domain',
      '@infra': './src/infra',
      '@main': './src/main',
      '@presentation': './src/presentation',
      '@tests': 'tests',
    },
  },
  resolve: {
    alias: {
      '@data': './src/data',
      '@domain': './src/domain',
      '@infra': './src/infra',
      '@main': './src/main',
      '@presentation': './src/presentation',
      '@tests': './src/tests',
    },
  },
});

```

**Comment by HarlemSquirrel** [comment]
Did you ever resolve this? I'm trying to set up TypeORM in a project that's been using vitest and am running into this issue.

It seems to happen when I initialize the data source with `AppDataSource.initialize()`. This error doesn't occur when running my express server, only when running vitest. I can run migrations from the CLI just fine. My tests run just fine when I comment out the `migrations` configuration in my data source.

#### Versions

typeorm@npm:0.3.19
vitest@npm:0.34.5

#### Code

```ts
// data-source.ts

export const AppDataSource = new DataSource({
  type: 'postgres',
  poolSize: 5,
  url: config.databaseUrl,
  logging: !config.isProduction,
  entities: [MyEntity],
  migrations: ['db/migrations/*.ts'], // Tests run when I comment out this line
  migrationsRun: false,
  subscribers: [],
});
```

```ts
import { it, describe, expect, beforeAll } from 'vitest';
import { AppDataSource } from '~db/data-source';

beforeAll(async () => {
  await AppDataSource.initialize()
    .then(() => {
      console.log('Data Source has been initialized!');
    })
    .catch((err) => {
      console.error('Error during Data Source initialization', err);
      throw err;
    });
});
```

tsconfig.json
```json
{
  "compilerOptions": {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "baseUrl": "./",
    "checkJs": false,
    "composite": true,
    "declaration": true,
    "declarationMap": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "incremental": true,
    "inlineSourceMap": true,
    "module": "nodenext",
    "moduleResolution": "nodenext",
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": false,
    "outDir": "./build",
    "preserveSymlinks": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "es2020",
    "tsBuildInfoFile": "./.tsbuildinfo",
    "useUnknownInCatchVariables": false,
    "paths": {
      "~app/*": ["./app/*"],
      "~config": ["./config"],
      "~config/*": ["./config/*"],
      "~db/*": ["./db/*"],
      "~integrations/*": ["./integrations/*"],
      "~shared/*": ["./shared/*"]
    }
  },
  "include": [
    "./**/*.ts",
    "vitest.config.mts",
    ".eslintrc.js"
  ],
  "exclude": ["build", "dist", "node_modules"]
}
```

```ts
// vitest.config.mts
const config = defineConfig({
  esbuild: {
    format: 'esm',
    target: 'es2020',
  },
  test: {
    testTimeout: 20000,
    exclude: ['build', 'node_modules'],
    globalSetup: [path.resolve(__dirname, './vitest.setup.ts')],
  },
  resolve: {
    alias: {
      '~app': path.resolve(__dirname, './app'),
      '~config': path.resolve(__dirname, './config'),
      '~db': path.resolve(__dirname, './db'),
      '~integrations': path.resolve(__dirname, './integrations'),
      '~shared': path.resolve(__dirname, './shared'),
      '~__recordings__': path.resolve(__dirname, './__recordings__'),
    },
  },
});
```

My single migration file so far
```ts
import { MigrationInterface, QueryRunner, Table, TableUnique } from 'typeorm';

export class AddCachedQueries1705597538034 implements MigrationInterface {
  public async up(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.createTable(
      new Table({
        // ...
```

**Reply by carlosmoliv** [reply]
I didn't, I decided to set aside that issue for now. Currently, every time I generate migrations, I convert the files from TS to JS.

**Reply by HarlemSquirrel** [reply]
I got around this by importing my migration and setting `migrations: [AddCachedQueries1705597538034]`  instead of a glob pattern.

**Reply by carlosmoliv** [reply]
Got it, I'm going to try this, thanks. The only disadvantage is the need to manually import each migration file, which sounds annoying. But if it fixes the issue, I can live with that little annoyance haha

**Reply by sheremet-va** [reply]
You can use glob import to make it easier: https://vitejs.dev/guide/features.html#glob-import

**Reply by gioboa** [reply]
I'm facing the same error `TypeError: Unknown file extension ".ts"` 
If I use `migrations` and `entities` with class names the system is working fine but I would like to have an automatic solution with the glob pattern.
the error is in [this typeorm method](https://github.com/typeorm/typeorm/blob/e7649d2746f907ff36b1efb600402dedd5f5a499/src/util/ImportUtils.ts#L11)
did you solve this issue?

**Reply by bgfernandes** [reply]
Having the same issue here as well, exactly as described. Good to know that doing `migrations` and `entities` with class names works, but it is a real pain with 50+ migrations, and also it's another step developers need to keep in mind when creating a new migration, to import it and add it to the list.

The glob import feature looks good, but I'm not using vite for building the project, it's a pure and simple `tsc`.

I wonder if there is a better solution out there.

**Reply by heldercostaa** [reply]
Exact same issue here, manually importing the entities and migrations worked, but still looking for a better solution

**Reply by zackpollard** [reply]
I just got this working using the previously mentioned vite glob-import as part of our globalSetup for some integration tests. I'll include the example of what I did for this, hopefully it can help some others :smile: 

We can use the import.meta.glob from vitest to dynamically import all the modules we need
`const modules = import.meta.glob('/src/migrations/*.ts', { eager: true });`

Then, for TS files, we can use the following mapper to convert those into classes that typeorm expects.
`Object.values(modules).map((module) => Object.values(module)[0])`

This can be passed directly to the `migrations` config property for a typeorm datasource.

Hope this helps some of you out!

**Comment by wqq20170106** [comment]
I also heve this questions.
![image](https://github.com/user-attachments/assets/5672044a-8ad5-46e2-9243-f0c9863c122a)
